scheduler/grpc_server.go

```diff
@@ -0,0 +1,71 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net"
+
+	pb "github.com/koraygocmen/gravitational/jobscheduler"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// server holds the GRPC scheduler server instance.
+type server struct{}
+
+// RegisterWorker registers a new worker on the server.
+// Workers call this method when they are coming online.
+func (s *server) RegisterWorker(ctx context.Context, r *pb.RegisterReq) (*pb.RegisterRes, error) {
+
+	workerID := newWorker(r.Address)
+
+	res := pb.RegisterRes{
+		WorkerID: workerID,
+	}
+
+	log.Printf("New worker with ID: %s is online\n", workerID)
+	return &res, nil
+}
+
+// DeregisterWorker deregisters an existing worker on the server.
+// Workers call this method when they are going offline.
+func (s *server) DeregisterWorker(ctx context.Context, r *pb.DeregisterReq) (*pb.DeregisterRes, error) {
+
+	delWorker(r.WorkerID)
```

Comment:
For a production system, you would probably have a real database backing your server. What happens if the server crashes here? In `newWorker` you allow overriding the worker, that's good, what do you think about attaching a TTL when a node registers/heartbeats itself? That way a node that simply stops heart beating automatically de-registers itself.

Reply:
I have thought about the heartbeat/ping etc for continuous monitoring of the online workers by the scheduler. I do believe that would be essential in a production system. I also addressed the necessity of a real database instead of the in-memory model in the main readme. But, I am going to add an additional comment here about the tradeoff and what would be required in a prod system. Thanks!